orca_detector/database_parser.py

```diff
@@ -247,63 +322,84 @@ def train_val_test_split(data_path=orca_params.DATA_PATH,
-    if dataset_type == DatasetType.TRAIN:
-        indices_file = os.path.join(data_path, 'train_map.p')
-    elif dataset_type == DatasetType.VALIDATE:
-        indices_file = os.path.join(data_path, 'validate_map.p')
-    elif dataset_type == DatasetType.TEST:
-        indices_file = os.path.join(data_path, 'test_map.p')
-    
-    if os.path.exists(indices_file):
-        with open(indices_file, 'rb') as f:
-            flattened = pickle.load(f)
-        print('\nLoaded {} dataset from {}'.format(dataset_type.name, indices_file))
+    features_file = os.path.join(data_path, dataset_type.name+'.features')
+
+    if os.path.exists(features_file):
+        with open(features_file, 'rb') as f:
+            features = pickle.load(f)
+        print('\nLoaded {} dataset from {}'.format(
+            dataset_type.name, features_file))
     else:
-        raise Exception('ERROR: run database_parser.py to generate datafiles first.')
-        
-    labels, files = zip(*flattened)
+        raise Exception(
+            'ERROR: run database_parser.py to generate datafiles first.')
 
-    return files, _onehot(labels)
+    labels, features = zip(*features)
 
+    features = np.array(features)
+    labels = np.array(labels)
     
+    # We need to remove one empty dimension
+    features = features[:, 0, :, :, :]
+
+    return np.array(features), np.array(labels)
+
+
+def create_label_encoding(classes, data_path=orca_params.DATA_PATH, save=True):
+    """
+        Saves LabelEncoder so inverse transforms can be recovered
+
+        Returns encoder : LabelEncoder
+    """
+
+    encoder = LabelEncoder()
+    encoder.fit(classes)
+    if save:
+        label_encoder_file = os.path.join(data_path, 'label_encoder.p')
+        # rename old files instead of overwriting.
+        _backup_datafile(label_encoder_file)
+        with open(label_encoder_file, 'wb') as fp:
+            pickle.dump(encoder, fp)
+            print('Saved label encoder to {}'.format(label_encoder_file))
+
+    return encoder
+
+
 if __name__ == '__main__':
 
-    # generate and save index files
-    train_val_test_split()
-    
-    # load all datasets to validate
-    for dataset_type in DatasetType:
-        files, labels = load_dataset(dataset_type=dataset_type)
-        print('\nDatasetType={}'.format(dataset_type.name))
-        print('Sample files:')
-        pprint.pprint(files[:5])
-        print('Counts by label:')
-        print(np.sum(labels, axis=0).astype(int))
-    
-    
\ No newline at end of file
+    # generate and save features
+    # read_files_and_extract_features()
+
+    # load the dataset features from disk.
+    train_features, train_labels = load_features(
+        orca_params.DATA_PATH, DatasetType.TRAIN)
+    validate_features, validate_labels = load_features(
+        orca_params.DATA_PATH, DatasetType.VALIDATE)
+
+    # shuffle, one hot and filter labels
+    classes = set(train_labels).union(set(validate_labels))
```

Comment:
Agreed. This is there is a TODO there, let's do a first iteration and decide how to reduce the classes. I'm not sure we need the other, it might be even better just to remove the excluded samples. One other thing that we need to eventually take care of, is the noise class. We'll talk about this on Monday.

Reply:
I think we definitely need an "Other" category.  The way softmax works, the predicted probability for all classes coming out of the classifier has to add to 1.  So if we don't have an "Other" class, we're forcing the model to always pick one of the desired classes for every audio sample it hears.  Since we have a bunch of samples that aren't the desired classes, I figured we should use them.

Since we already have the logic from the previous version I don't see why we'd remove that code completely now, knowing that we'll need to add it back in before we start training models.